tests: gdbstub: simple test

golowanow opened the pull request:
Gdbstub test with Twister pytest plugin:
* clone samples/subsys/debug/gdbstub to tests and convert it back to a build-only sample aligned with documentation.
* some improvements with pytest fixtures, parametrization, and expected pattern matching on outputs from GDB and the test application.

as suggested at https://github.com/zephyrproject-rtos/zephyr/pull/63472#pullrequestreview-1660080073
and follows #63061

golowanow committed:
tests: samples: gdbstub: make test out of the sample

Clone samples/subsys/debug/gdbstub to tests and convert it back
to a build-only sample aligned with documentation.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
tests: gdbstub: Improve test case

Gdbstub test improvements: using pytest fixtures, parametrization, and
expected pattern matching on outputs from GDB and the test application.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
config: qemu: Add QEMU_EXTRA_FLAGS config option

Add QEMU_EXTRA_FLAGS as QEMU board config option.

This allows Twister tests to provide additional device setup
commands to QEMU in prj.conf or testcase.yaml configuration files.

Example use case: to setup TCP or UDP network interfaces
with non-conflicting port numbers in different test suites
to avoid conflicts when Twister run tests in parallel on the
same host.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
tests: gdbstub: Add GDB remote target parameter

Add `gdb_target_remote` test parameter for GDB `target remote`
command instead of its hardcoded value to allow different types
of gdbstub serial interfaces as well as different TCP ports in
gdbstub test suites possibly run in parallel on the same host.

Move all GDB log configuration parameters from GDB script to
the fixture code.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
tests: gdbstub: qemu: Cross-validate test behavior

Add a testcase to run the same test application and GDB script
which we use for Zephyr GDB stub testing, but now with
the GDB stub enabled at QEMU itself using it as a reference
RDP backend implementation. This allows to check the Zephyr's
gdbstub implementation has similar behavior as the reference.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

nashif commented on tests/subsys/debug/gdbstub/prj.conf:
```diff
@@ -1,5 +1,3 @@
-CONFIG_GDBSTUB=y
```
this test is about testing the gdbstub feature in zephyr, so in what cases would this not be needed?

golowanow replied:
> this test is about testing the gdbstub feature in zephyr, so in what cases would this not be needed?

it is not needed for Qemu reference test case.

golowanow replied:
changed to have common config options for gdbstub to be enabled for this test suite

nashif commented on tests/subsys/debug/gdbstub/testcase.yaml:
```diff
@@ -17,12 +18,40 @@ tests:
+
+  # Connect to QEMU gdbstub and run the same simple GDB script
+  # Use non-default QEMU gdbstub port 1235
+  debug.gdbstub_qemu.breakpoints:
```
how does this scenario test the gdbn stub feature in zephyr? It seems to be testing general gdb connection here.

golowanow replied:
This test suite runs the same GDB script on two gdbstub implementations: our main target Zephyr gdbstub, and Qemu gdbstub reference target.
The latter is not only a part of Zephyr debugging toolset, but here serves as a reference implementation of RDP backend, so we check the expected behavior is the same.

nashif replied:
ok

nashif reviewed (changes requested):
looks good overall, just a few comments and clarifying questions.

nashif commented:
@gchwier  @gopiotr  @PerMac  please review

golowanow commented on tests/subsys/debug/gdbstub/testcase.yaml:
```diff
@@ -5,7 +5,8 @@
 tests:
-  debug.gdbstub.breakpoints:
+  # Connect to Zephyr gdbstub and run a simple GDB script
+  debug.gdbstub_zephyr.breakpoints:
```
this test is to differ Zephyr gdbstub test from QEMU gdbstub.

golowanow replied:
changed to `debug.gdbstub.breakpoints`

golowanow commented on samples/subsys/debug/gdbstub/sample.yaml:
```diff
@@ -8,8 +8,8 @@ sample:
   name: gdbstub sample
 tests:
```
just to double check our intentions in regard of the gdbstub sample:
to me it looks like we need to keep it as an essential part of [gdbstub documentation](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/services/debugging/gdbstub.rst#id50) assuming the gdbstub is a developer's tool with an example how to use it 'manually' without any relation to testing and according to [sample vs test](https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html).
Also, this PR's test has extra pytest-specific details and adjusted to automated execution, so extra details might be confusing.

nashif replied:
right, but now we are back to a sample not being tested.

golowanow replied:
> right, but now we are back to a sample not being tested.

not sure I understand your point: do we need to test the sample or the gdbstub feature ?

nashif replied:
both

golowanow replied:
the gdbstub sample removed in the last commit added to this PR.

golowanow commented on tests/subsys/debug/gdbstub/pytest/test_gdbstub.py:
```diff
@@ -4,31 +4,67 @@
+#
+
+@pytest.fixture(scope="module")
+def expected_app():
```
> this is all related to one test, why do we need a fixture? They can be defined in the test_ itself.

just to follow the pytest way of 'fixturizing' test data from test logic.

gchwier replied:
It is always good to have separated data from the logic, so the idea is good. Here it is only one test, so you can use just a method instead of fixture (however fixture is also ok)

golowanow commented on tests/subsys/debug/gdbstub/testcase.yaml:
```diff
@@ -1,13 +1,18 @@
     tags:
       - debug
       - gdbstub
+      - pytest
```
agree, I'll remove it

golowanow replied:
removed

golowanow commented on scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures.py:
```diff
@@ -66,3 +68,11 @@ def is_mcumgr_available() -> None:
+
+
+@pytest.fixture(scope='function')
+def gdb_batch(dut: DeviceAdapter, gdb_target_remote: str, gdb_script: str, gdb_timeout) -> CompletedProcess:
```
> This fixture cannot be here (in pytst-twister-harness plugin). It uses local fixtures, defined in test dir (options added in conftest.py)

yes, it might be a problem to depend on some external/local fixtures.

@gchwier, what might be a better way for pytest plugin to use command line parameters here ?

gchwier replied:
For now the easiest solution is to keep that fixture in conftest.py in local test dir.
But if we want to have it available for other tests, than we can add command line parameters to [pytest-twister-harness/src/twister_harness/plugin.py](https://github.com/zephyrproject-rtos/zephyr/blob/f6185221440b7fe9f0c1b17702c8231105edc06b/scripts/pylib/pytest-twister-harness/src/twister_harness/plugin.py#L21)
and grab them in a fixture:
```
@pytest.fixture(scope='function')
def gdb_batch(dut: DeviceAdapter, request: pytest.FixtureRequest) -> CompletedProcess:
    gdb_target_remote = request.config.getoption('--gdb-target-remote')
    gdb_script = request.config.getoption('--gdb-script')
    gdb_timeout = request.config.getoption('--gdb-timeout')
    gdb = GDBrunner(dut)
    return gdb.run_script(gdb_target_remote, gdb_script, gdb_timeout)
```
Could be fine also to check, if gdb runner is available etc. and then use `pytest.skip(with reason)`

golowanow replied:
> This fixture cannot be here (in pytst-twister-harness plugin). It uses local fixtures, defined in test dir (options added in conftest.py)

changed to be a factory method fixture

gchwier replied:
> changed to be a factory method fixture

IMHO it is not better solution. Why do you keep that fixture in the plugin, if you need add another fixture in local conftest (to just pass parameters)?
Better to keep `gdb.py` in the plugin helpers methods, and in local conftest add a fixture, that takes given parameters and calls `GDBrunner`.
Or to add parameters to `plugin.py` and keep gdb fixture in the plugin (then please update `pytest.rst` docs, there is a section with available fixtures: https://docs.zephyrproject.org/latest/develop/test/pytest.html#helpers-fixtures )

golowanow replied:
changed back to have gdb fixture local to the test, so no changes at the pytest plugin in this PR.

golowanow commented on scripts/pylib/pytest-twister-harness/src/twister_harness/helpers/gdb.py:
```diff
@@ -0,0 +1,64 @@
+logger = logging.getLogger(__name__)
+
+
+class GDBrunner:
```
`dgbrunner` fixture's promotion to plugin level is the last commit in this PR, so it might be separated to another PR.

golowanow replied:
> we must find a good way, how to import cmake_cache package from twisterlib, how to get extra parameters etc.

i've also thought about cmake_cache as another pytest plugin's future fixture

golowanow replied:
outdated as removed from this PR

golowanow committed:
samples: gdbstub: doc: Remove redundant sample

Remove gdbstub sample (samples/subsys/debug/gdbstub) as duplicated
by a test (tests/subsys/debug/gdbstub).

Update the GDB stub documentation.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
> looks good overall, just a few comments and clarifying questions.

@nashif could you please look again after the changes done here.
